abnf/odata-aggregation-abnf.txt

```diff
@@ -132,13 +132,20 @@ pathSegment     = ( complexProperty / complexColProperty ) [ "/" qualifiedComple
 
 groupingProperty = grpPropPathPrefix
                    ( entityNavigationProperty [ "/" qualifiedEntityTypeName ]
-                   / primitiveProperty 
+                   / entityAnnotationInQuery  [ "/" qualifiedEntityTypeName ]
+                   / primitiveProperty        [ "/" grpPropAnnoSegment ]
                    / complexProperty
+                   / grpPropAnnoSegment
                    )
 grpPropPathPrefix  = [ ( qualifiedEntityTypeName / qualifiedComplexTypeName ) "/" ] *( grpPropPathSegment "/" )
 grpPropPathSegment = complexProperty          [ "/" qualifiedComplexTypeName ]
                    / entityNavigationProperty [ "/" qualifiedEntityTypeName  ]

```

Comment:
Why not use the same pattern as in the ABNF for `$select` and `$expand`: https://github.com/oasis-tcs/odata-abnf/blob/ba76fc36afc38cf7a426587f8024987b7e44a9af/abnf/odata-abnf-construction-rules.txt#L286

```suggestion
                   / ( entityNavigationProperty / entityAnnotationInQuery ) [ "/" qualifiedEntityTypeName  ]
```

Reply:
Because
```
entityAnnotationInQuery  [ "/" qualifiedEntityTypeName  ]
```
can also occur at the end of the `groupingProperty` rule. Therefore I made it part of the `grpPropAnnoSegment` segment rule, which can appear both in the middle and at the end.

Reply:
Grouping by entities means grouping is by entity-id, and a type-cast does not change the entity-id, so: what effect will a type-cast at the end of a grouping property have for an entity? 

Which also raises the question how grouping by entity-valued annotation will work: what is the entity-id to group by?

Reply:
A type-cast is never the last segment, it is needed only if a (primitive) property from the casted type follows.

An entity-valued instance annotation must have an entity-id (otherwise it's not an entity).

Reply:
> a type-cast and the end is not needed, but it is allowed today

We should disallow it then.